tests/unit/ai/test_retry.py

```
@@ -0,0 +1,169 @@
+    fn()
+    delays = [call.args[0] for call in mock_sleep.call_args_list]
+    for delay in delays:
+        assert_that(delay).is_less_than_or_equal_to(25.0)
+
+
+def test_retry_does_not_retry_non_ai_exceptions():
+    """Verify non-AI exceptions propagate immediately without retrying."""
+    call_count = 0
+
+    @with_retry(max_retries=3)
+    def fn():
```

Ruff (0.15.6)

[warning] Missing return type annotation for private function `fn`

Add return type annotation: `Never`

(ANN202)